File: serverban.go
package unrealircd

import (
	"errors"
)

// ServerBan handles server ban operations
type ServerBan struct {
	querier Querier
}

// Add adds a ban
func (sb *ServerBan) Add(name, banType, duration, reason string) (interface{}, error) {
	result, err := sb.querier.Query("server_ban.add", map[string]interface{}{
		"name":             name,
		"type":             banType,
		"reason":           reason,
		"duration_string":  duration,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// Delete deletes a ban
func (sb *ServerBan) Delete(name, banType string) (interface{}, error) {
	result, err := sb.querier.Query("server_ban.del", map[string]interface{}{
		"name": name,
		"type": banType,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil
}

// GetAll returns a list of all bans
func (sb *ServerBan) GetAll() (interface{}, error) {
	result, err := sb.querier.Query("server_ban.list", nil, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if list, ok := res["list"]; ok {
			return list, nil
		}
	}

	return nil, errors.New("Invalid JSON Response from UnrealIRCd RPC")
}

// Get gets a specific ban
func (sb *ServerBan) Get(name, banType string) (interface{}, error) {
	result, err := sb.querier.Query("server_ban.get", map[string]interface{}{
		"name": name,
		"type": banType,
	}, false)
	if err != nil {
		return nil, err
	}

	if res, ok := result.(map[string]interface{}); ok {
		if tkl, ok := res["tkl"]; ok {
			return tkl, nil
		}
	}

	return nil, nil // not found
}

File: UnrealIRCdRPC/ServerBan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Text.Json;
using UnrealIRCdRPC.Models;

namespace UnrealIRCdRPC
{
    // ServerBan handles server ban operations
    public class ServerBan
    {
        private readonly IQuerier _querier;

        public ServerBan(IQuerier querier)
        {
            _querier = querier;
        }

        // Add adds a ban
        public async Task<TklInfo?> AddAsync(string name, string banType, string duration, string reason)
        {
            var result = await _querier.QueryAsync("server_ban.add", new
            {
                name,
                type = banType,
                reason,
                duration_string = duration
            }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                // Convert the JsonElement to TklInfo using JSON deserialization
                return tklElement.Deserialize<TklInfo>();
            }
            return null;
        }

        // Delete deletes a ban
        public async Task<JsonElement?> DeleteAsync(string name, string banType)
        {
            var result = await _querier.QueryAsync("server_ban.del", new { name, type = banType }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                return tklElement;
            }
            return null;
        }

        // GetAll returns a list of all bans
        public async Task<IReadOnlyList<string>> GetAllAsync()
        {
            var result = await _querier.QueryAsync("server_ban.list", null, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object)
            {
                if (result.Value.TryGetProperty("list", out var listElement) && listElement.ValueKind == JsonValueKind.Array)
                {
                    var banNames = new List<string>();
                    foreach (var banElement in listElement.EnumerateArray())
                    {
                        if (banElement.ValueKind == JsonValueKind.Object &&
                            banElement.TryGetProperty("name", out var nameElement) &&
                            nameElement.ValueKind == JsonValueKind.String)
                        {
                            banNames.Add(nameElement.GetString()!);
                        }
                    }
                    return banNames;
                }
            }
            throw new Exception("Invalid JSON Response from UnrealIRCd RPC");
        }

        // Get gets a specific ban
        public async Task<TklInfo?> GetAsync(string name, string banType)
        {
            var result = await _querier.QueryAsync("server_ban.get", new { name, type = banType }, false);
            if (result.HasValue && result.Value.ValueKind == JsonValueKind.Object &&
                result.Value.TryGetProperty("tkl", out var tklElement))
            {
                // Convert the JsonElement to TklInfo using JSON deserialization
                return tklElement.Deserialize<TklInfo>();
            }
            return null; // not found
        }
    }
}